repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +90,53 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
   def execute(code: String): Int = {
     val statementId = newStatementId.getAndIncrement()
     _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+
     Future {
-      _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      setJobGroup(statementId)
+      _statements(statementId).state.compareAndSet(StatementState.Waiting, StatementState.Running)
 
-      _statements(statementId) =
-        new Statement(statementId, StatementState.Available, executeCode(statementId, code))
+      val executeResult = if (_statements(statementId).state.get() != StatementState.Cancelling) {
+        executeCode(statementId, code)
+      } else {
+        null
+      }
+
+      _statements(statementId).output = executeResult
+      _statements(statementId).state.compareAndSet(StatementState.Running, StatementState.Available)
+      _statements(statementId).state.compareAndSet(
+        StatementState.Cancelling, StatementState.Cancelled)
     }
+
     statementId
   }
 
+  def cancel(statementId: Int): Unit = {
+    if (_statements.contains(statementId)) {
+      val oldStmtState = _statements(statementId).state.getAndSet(StatementState.Cancelling)
+
+      info(s"Statement $statementId is cancelling")
+
+      Future {
+        if (oldStmtState == StatementState.Running) {
+          val currTime = System.currentTimeMillis()
+          while (_statements(statementId).state.get() != StatementState.Cancelled &&
+            System.currentTimeMillis() - currTime <= TimeUnit.SECONDS.toMillis(30)) {
+            _sc.foreach(_.cancelJobGroup(statementId.toString))
+            Thread.sleep(livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TRIGGER_INTERVAL))
+          }
+        } else if (oldStmtState == StatementState.Available) {
+          _statements(statementId).state.set(StatementState.Cancelled)
+        }
+      }(cancelExecutor)
+    }
+  }
+
   def close(): Unit = {
     executor.shutdown()
     interpreter.close()
   }
 
-  def clearStatements(): Unit = synchronized {
+  def clearStatements(): Unit = {
```

Comment:
`cancel()` doesn't really work if `_statements.clear()` is called while it's running.
We can fix this with a lock or just remove `clearStatements()` since no one is using it.